src/persist-client/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+//   has some sort of encode/decode API that uses byte slices, Buf+BufMut, or
+//   proto, depending on what we decide.
+// - PhantomData is a placeholder for internal state that users don't care
+//   about.
+
+// WIP
+// - Which of these need to be durably storable (for restarts) or transmittable
+//   over the network? The formalism document describes certain things as having
+//   this property, but does this translate to a persist requirement as well?
+//   Candidates: Location, ReadHandle, WriteHandle, others? Definitely: Id,
+//   SnapshotShard.
+// - Same question for clone-able. The formalism document describes the
+//   storage-level read and write capabilities as being clone-able, but does
+//   this translate to a persist requirement as well?
+// - Figure out the contract of tokio Runtimes and who is responsible for making
+//   sure they're in the TLC.
+// - I've split up getting a snapshot vs tailing for new changes below, because
+//   I think it simplifies the story about multi-process snapshots. We could
+//   also instead do something more analogous to SubscribeAt from the formalism
+//   doc.
+// - We'll want to expose overall collection frontier information (since and
+//   upper) at the very least for introspection. There's a few options here so
+//   we should think about how this will be used.
+// - Leases for read and write handles, as described in formalism.
+// - I think some things (readers at least?) will want to connect to the persist
+//   controller over rpc. At the very least, this is something that we may want
+//   to hook up for latency optimizations (making inputs to the state machine
+//   push instead of pull). How does this information get plumbed around? Maybe
+//   a dns name that cloud is responsible for making sure points at the right
+//   thing?
+// - To what extent should this adopt the naming conventions for STORAGE used in
+//   the formalism doc? Many of them have direct analogues down at the persist
+//   level.
+// - At the moment, we have both a STORAGE collection concept as well as a
+//   persist collection concept. One of these should be renamed :).
+
+/// A location in s3, other cloud storage, or otherwise "durable storage" used
+/// by persist. This location can contain any number of persist collections.
+///
+/// WIP: This also needs etcd (or whatever we pick) connection information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Location {
+    bucket: String,
+    prefix: String,
+}
+
+/// An opaque identifier for a persist collection.
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub struct Id([u8; 16]);
+
+pub struct Client {
+    _phantom: PhantomData<()>,
+}
+
+impl Client {
+    pub async fn new(location: Location) -> Result<Self, Error> {
+        todo!("{:?}", location)
+    }
+
+    pub async fn new_collection<K, V, T, D>(
+        &self,
+    ) -> Result<(ReadHandle<K, V, T, D>, WriteHandle<K, V, T, D>), Error>
```

Comment:
This should either accept or return a collection id, right?

Reply:
at the moment i have it assigning a random one, but this is exactly the sort of thing i'm looking to hash out in this PR. I think your comment in the other thread about the collection id being orphaned is a good one.

Q back at you: if this takes an id, then the impl is basically exactly the same as open_collection, except that this one can potentially guarantee that the id isn't already in use and that the since and uppers of the returned handles are both `Antichain::from_elem(T::minimum())`. do you think the distinction is worth keeping? persist can do whatever here, it's really just what's best for storage